storage/innobase/row/row0ins.cc

```diff
@@ -1755,18 +1755,57 @@ row_ins_check_foreign_constraint(
 		cmp = cmp_dtuple_rec(entry, rec, offsets);
 
 		if (cmp == 0) {
-			if (rec_get_deleted_flag(rec,
-						 rec_offs_comp(offsets))) {
-				err = row_ins_set_shared_rec_lock(
-					LOCK_ORDINARY, block,
-					rec, check_index, offsets, thr);
-				switch (err) {
-				case DB_SUCCESS_LOCKED_REC:
-				case DB_SUCCESS:
-					break;
-				default:
-					goto end_scan;
-				}
+                  if (DICT_TF2_FLAG_IS_SET(check_index->table,
+                                           DICT_TF2_VERSIONED)) {
+                    mem_heap_t *clust_heap = mem_heap_create(256);
+                    ulint clust_offsets_[REC_OFFS_NORMAL_SIZE];
+                    ulint *clust_offsets = clust_offsets_;
+                    rec_offs_init(clust_offsets_);
+                    btr_pcur_t clust_pcur;
+
+                    ut_a(!dict_index_is_clust(check_index));
+                    dict_index_t *clust_index =
+                        dict_table_get_first_index(check_index->table);
+
+                    dtuple_t *ref = row_build_row_ref(
+                        ROW_COPY_POINTERS, check_index, rec, clust_heap);
+
+                    mtr_t clust_mtr;
+                    mtr_start(&clust_mtr);
+                    btr_pcur_open(clust_index, ref, PAGE_CUR_GE,
+                                  BTR_SEARCH_LEAF, &clust_pcur, &clust_mtr);
```

Comment:
In case `check_index` is clustered (i.e. main index) there is need to open second cursor. Use previously opened one.